core/src/main/scala/org/broadinstitute/dsde/rawls/snapshot/SnapshotReferenceCreationValidator.scala

```diff
@@ -0,0 +1,89 @@
+  private def isWorkspaceProtected: Boolean = workspaceContext.bucketName.startsWith("fc-secure")
+
+  private def isSnapshotProtected: Boolean = snapshot.getSource.asScala.exists(_.getDataset.isSecureMonitoringEnabled)
+
+  private def snapshotPlatform: SnapshotCloudPlatform = {
+    val snapshotSourceDatasets = snapshot.getSource.asScala
+    val datasetCloudPlatforms = snapshotSourceDatasets.map(_.getDataset.getCloudPlatform).filter(_ != null).toSet
+
+    // assume GCP if datasets don't come from a source explicitly marked with another platform
+    if (datasetCloudPlatforms.isEmpty) {
+      return SnapshotCloudPlatform.GCP
+    }
+
+    // fail if datasets sources somehow span multiple clouds
+    if (datasetCloudPlatforms.size > 1) {
+      throw new IllegalStateException(
+        s"snapshot ${snapshot.getId} unexpectedly has more than one cloudPlatform: $datasetCloudPlatforms"
+      )
+    }
+
+    // fail if dataset sourced from something other than GCP aren't all explicitly marked as such
+    // as this might indicate co-mingled GCP (blank) & non-GCP datasets
+    val onlyCloudPlatform = datasetCloudPlatforms.take(1).last
+    onlyCloudPlatform match {
+      case SnapshotCloudPlatform.GCP => onlyCloudPlatform
+      case _ =>
+        if (snapshotSourceDatasets.forall(_.getDataset.getCloudPlatform == onlyCloudPlatform)) onlyCloudPlatform
+        else
+          throw new IllegalStateException(
+            s"snapshot ${snapshot.getId} unexpectedly has sources from more than one cloudPlatform"
+          )
+    }
+
+  }
+
+  // Ideally this would rely on Terra Policy Service, but until TPS is enabled for GCP
+  // We'll have to use this workaround for identifying protected status.
+  @throws(classOf[ProtectedDataException])
+  def validateProtectedStatus(): Unit =
+    if (!isWorkspaceProtected && isSnapshotProtected) {
+      throw new ProtectedDataException("Unable to add protected snapshot to unprotected workspace.")
+    }
+
+  @throws(classOf[UnsupportedPlatformException])
+  @throws(classOf[PlatformBoundaryException])
+  def validatePlatforms(workspacePlatform: WorkspaceCloudPlatform): Unit = {
+    // Defining acceptable combinations is overkill when all we really need to do is prevent anything Azure, but it's
+    // here as a safeguard against us introducing support for Azure without deliberately addressing the issue of
+    // snapshots by ref across cloud platforms.
+    val acceptableCombinations = Map((SnapshotCloudPlatform.GCP, WorkspaceCloudPlatform.Gcp))
+
+    if (!acceptableCombinations.get(snapshotPlatform).contains(workspacePlatform)) {
+      throw new PlatformBoundaryException(
+        "Snapshots by reference are not supported across the given cloud boundaries " +
+          s"(snapshot: ${snapshotPlatform.toString.toLowerCase}, workspace: ${workspacePlatform.toString.toLowerCase})."
+      )
+    }
```

Comment:
one of my favorite features of Scala is its rich pattern-matching. Again, the code you have is fine but to be Scalalike:
```scala
    (snapshotPlatform, workspacePlatform) match {
      case (SnapshotCloudPlatform.GCP, WorkspaceCloudPlatform.Gcp) => // ok
      case _ =>
        throw new PlatformBoundaryException(
          "Snapshots by reference are not supported across the given cloud boundaries " +
            s"(snapshot: ${snapshotPlatform.toString.toLowerCase}, workspace: ${workspacePlatform.toString.toLowerCase})."
        )
```